### PR TITLE
Docs/4277 add doc partial partial

### LINKED ISSUE #4277: Document resource sets and partial compiles
Tasks:

- [x] Document the concepts of a resource set and a partial compile.
- [x] Document the limitations (e.g. index checks cannot be done completely).
- [ ] Add changelog entry regarding this new feature

# Update 2022-08-05:
Moving back to next because Florent will be off for the next week. Remaining tasks:
- [ ] process review comments on #4600
- [ ] Add changelog entry regarding this new feature
- [ ] work with Pouriya to improve documentation quality
- [ ] final review

### PR DESCRIPTION
**Pull request by sanderr**
# Description

Rework of #4600 in cooperation with Pouriya ([diff since last review](https://github.com/inmanta/inmanta-core/pull/4766/files/9a90487bdc592f88febf93dc8359c2dae3e249d8..HEAD?diff=split&w=0) (hope this link works)).

closes #4277 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: docs/resource_sets.rst [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+Limitations
+*************
+
+* The compiler cannot verify all constraints that would be verified when a full build is run. Some index constraints, for instance, cannot be verified. The model creator is in charge of making sure that these constraints are met.
+* If just a partial compile is performed, it is possible for a shared resource to become obsolete because numerous resources from distinct resource sets may depend on the same shared resource (not associated with a specific resource set).
+  The shared resource will become obsolete when a partial compile deletes the last resource that depended on it, but it is preserved as a server-managed resource because partial compiles cannot delete shared resources.
+  A full compile is required to remove shared resources. Scheduled complete compilations that ``garbage-collect`` these shared resources are one way to fix this.
+  The :inmanta.environment-settings:setting:`auto_full_compile` environment setting is used to schedule full compilations.
+  As an example, to plan a daily full compile for 01:00 UTC, use the ``auto full compile`` environment setting:  "0 1 * * *".
```
Comment (FloLey) [inline docs/resource_sets.rst:93]: I don't like the formatting of the cron expression but it don't know what the conventions are so maybe no change is needed.

**Review by FloLey — APPROVED** [review]
I think this version is already a lot more clearer for someone who doesn't know anything about resource sets than the one I had started :p

File: docs/resource_sets.rst (hunk 1 above)
Comment (sanderr) [reply docs/resource_sets.rst:93]: 1. It's already been released so there's little we can do at this point.
2. It really just follows the cron format so while it's not the most intuitive it is familiar to many.

File: docs/resource_sets.rst [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+
+Limitations
+*************
+
+* The compiler cannot verify all constraints that would be verified when a full build is run. Some index constraints, for instance, cannot be verified. The model creator is in charge of making sure that these constraints are met.
+* If just a partial compile is performed, it is possible for a shared resource to become obsolete because numerous resources from distinct resource sets may depend on the same shared resource (not associated with a specific resource set).
+  The shared resource will become obsolete when a partial compile deletes the last resource that depended on it, but it is preserved as a server-managed resource because partial compiles cannot delete shared resources.
+  A full compile is required to remove shared resources. Scheduled complete compilations that ``garbage-collect`` these shared resources are one way to fix this.
```
Comment (arnaudsjs) [inline docs/resource_sets.rst:91]:
```suggestion
  A full compile is required to remove shared resources. Scheduled full compilations that ``garbage-collect`` these shared resources are one way to fix this.
```

File: docs/resource_sets.rst (hunk 1 above)
Comment (arnaudsjs) [inline docs/resource_sets.rst:93]:
```suggestion
  As an example, to plan a daily full compile for 01:00 UTC, use the ``auto_full_compile`` environment setting:  ``0 1 * * *``.
```

File: tests/test_docs_snippets.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        Add handlers for the base model and set up the snippetcompiler.
+        """
+        handlers_addition: str = f"""
+            import {handler_module_name} as handler
+
+            Router.agent [1] -- std::AgentConfig
+            implementation bind_agent for Router:
+                self.agent = std::AgentConfig[agentname="router_agent"]
+            end
+            implement Router using bind_agent
```
Comment (arnaudsjs) [inline tests/test_docs_snippets.py:89]: Let's add a comment regarding why this should be added.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]
I see in the output of the compilation of our documentation that our sphinx extensions is unable to lex the Inmanta code in the examples. I am not saying we should fix that immediately, but we should at least create a ticket to get this fixed. It's always nice to have proper syntax highlighting in our code snippets.

File: docs/resource_sets.rst (hunk 2 above)
Comment (sanderr) [reply docs/resource_sets.rst:91]: Ah, missed one

**Comment by sanderr** [comment]
inmanta/inmanta-sphinx#54

**Review by arnaudsjs — APPROVED** [review]

File: docs/resource_sets.rst (hunk 1 above)
Comment (FloLey) [reply docs/resource_sets.rst]: I meant the how it is displayed as 'normal' text in the doc, the layout. I don't have an issue with the cron format in itself :p

Comment (sanderr) [reply docs/resource_sets.rst]: Ah, I see, that we did change because Arnaud requested the same.

**Pull request closed without merging**